tools/wpt/browser.py

```diff
@@ -516,264 +516,201 @@ def version(self, binary=None, webdriver_binary=None):
 
-        return "%s%s" % (platform, bits)
+        return f"{self.platform.lower()}{bits}"
 
     def _chromium_platform_string(self):
-        platform = self.platforms.get(uname[0])
-
-        if platform is None:
-            raise ValueError("Unable to construct a valid Chromium package name for current platform")
+        if self.platform is None:
+            raise ValueError(
+                "Unable to construct a valid Chromium package name for current platform")
 
-        if (platform == "Linux" or platform == "Win") and uname[4] == "x86_64":
-            platform += "_x64"
+        if (self.platform == "Linux" or self.platform == "Win") and uname[4] == "x86_64":
+            return f"{self.platform}_x64"
 
-        return platform
+        return self.platform
 
     def _chromium_package_name(self):
-        return "chrome-%s" % self.platforms.get(uname[0]).lower()
+        return f"chrome-{self.platform.lower()}"
 
-    def _latest_chromium_snapshot_url(self):
-        # Make sure we use the same revision in an invocation.
+    def _format_chromium_snapshot_url(self, revision=None):
         architecture = self._chromium_platform_string()
-        if self._last_change is None:
-            revision_url = "https://storage.googleapis.com/chromium-browser-snapshots/%s/LAST_CHANGE" % architecture
-            self._last_change = get(revision_url).text.strip()
-        return "https://storage.googleapis.com/chromium-browser-snapshots/%s/%s/" % (architecture, self._last_change)
+        # Make sure we use the same revision in an invocation.
+        # If we have a last revision used, that revision takes priority.
+        if self.last_revision_used is None and revision is None:
+            # No revision provided or previously used. use latest revision.
+            self.revision_attempted = self._get_latest_chromium_revision(architecture)
+        elif revision is not None:
+            self.revision_attempted = revision
+        else:
+            self.revision_attempted = self.last_revision_used
 
-    def find_nightly_binary(self, dest):
-        if uname[0] == "Darwin":
-            return find_executable("Chromium",
-                                   os.path.join(dest, self._chromium_package_name(), "Chromium.app", "Contents", "MacOS"))
-        # find_executable will add .exe on Windows automatically.
-        return find_executable("chrome", os.path.join(dest, self._chromium_package_name()))
+        # The attempted revision is kept. If it is successfully used,
+        # it will be used again if another component is also installed (browser/webdriver/mojojs).
+        return ("https://storage.googleapis.com/chromium-browser-snapshots/"
+                f"{architecture}/{self.revision_attempted}/")
 
-    def find_binary(self, venv_path=None, channel=None):
-        if channel == "nightly":
-            return self.find_nightly_binary(self._get_dest(venv_path, channel))
+    def _get_chromium_url_by_version(self, version=None):
```

Comment:
This is called without a version most of the time. Can you add some documentation explaining what this is for and how it's intended to be used? Because it seems a bit odd, if we have a version number, then why do we need a revision? I think I know the answer in some cases, but it'd be good to write it down.

Reply:
That's a good idea. Where is the best place to document this? Just as a comment in the code, or somewhere else? Also, since it is true that you're more likely to _not_ pass the version, this might be better named as something like `_get_chromium_download_url` to feel less version-reliant.

Reply:
Additionally, I reimplemented this to not keep revisions attempted and successfully used as attributes, and instead just keep the url that was successfully used for reuse.
https://github.com/web-platform-tests/wpt/blob/615e384f4863ee297d4fc6afde94f02cb5c90b7b/tools/wpt/browser.py#L732-L736
Hopefully this is more intuitive.